lib/temporal/activity/context.rb

```diff
@@ -30,9 +39,30 @@ def async_token
         )
       end
 
+      def timed_out?
+        !start_to_close_deadline.nil? && Time.now > start_to_close_deadline
+      end
+
       def heartbeat(details = nil)
```

Comment:
Not exactly related, but since you're adding `ActivityTimedOut` it may be an opportune scope creep... Another thing we want to do here is detect if the worker is draining, i.e. if `Temporal::Activity::Poller.shutting_down?` is true, and raise if it is.  (I'll send you a link to the JIRA internally)  One idea is to have the poller set a thread-local under a mutex that we can check here.

Reply:
I was thinking a bit about this while coding this up. I was considering a more complicated solution originally involving raising a special error on each heartbeat thread but checking the centralized `shutting_down?` bit on the poller may be sufficient and simpler. I'll see how complicated this might be before deciding to include it in this PR or to break it out into a follow up.